### DISCUSSION TITLE
VP8 & Unity Render Streaming support

### DISCUSSION THREAD

**Discussion by felchs** [paullouisageneau/libdatachannel#567] [Q&A]
Hello,

I found this amazing lib but the VP8 support is not availble ok?
I was trying to connect to Unity and I got an error: 
`[impolite-Unity.RenderStreaming.PeerConnection] Failed to set local answer sdp: Failed to set local video description recv parameters for m-section with mid='video-stream'.`

I researched it and I found out that unity only supports VP8 software decoding.

So I need to send the packets via VP8 ok?

**Comment by paullouisageneau** [comment]
Hello,

You can forward an existing VP8 RTP stream with the library, but you can't packetize VP8 frames into RTP with it for now (There no VP8 packetizer, only a H264 packetizer, see #558).

The errors seems to be that Unity does not support the track parameters,you should open an issue with more details (Track setup, SDP offer, etc).

**Reply by felchs** [reply]
The signaling process is also differs from the StreamedVideo from unity https://docs.unity3d.com/Packages/com.unity.renderstreaming@3.1/manual/video-streaming.html, I'm having to adjust something on the protocols.

I'll open an issue.

Is there an example of  opening and exchanging a VP8 RTP stream without packetizer?
Is the sfu-media one I could modify adding:

https://github.com/paullouisageneau/libdatachannel/blob/master/examples/sfu-media/main.cpp line 55:
addVP8Codec(96) instead of media.addH264Codec(96) ?

I'm kinda new to this technology what's a packetizer? what it does?

As as a workaround, I was not able to send video stream so I was trying to send pure bytes data over Data channel....

**Comment by paullouisageneau** [comment]
> I'm kinda new to this technology what's a packetizer? what it does?

The packetizer packages raw frames (for instance what you get from a webcam) into RTP packets.

> Is there an example of opening and exchanging a VP8 RTP stream without packetizer?
> Is the sfu-media one I could modify adding:
>
> https://github.com/paullouisageneau/libdatachannel/blob/master/examples/sfu-media/main.cpp line 55:
addVP8Codec(96) instead of media.addH264Codec(96) ?

Yes indeed, the sfu-media is pretty close to direct RTP streaming.

**Reply by felchs** [reply]
You said that "The packetizer packages raw frames (for instance what you get from a webcam) into RTP packets."

The question I have now is: with direct RTP streaming what kind of packet is transmitted? it's the stream file itself without packetizer ok, but the difference is that only raw data is transmitted not frames?

**Reply by paullouisageneau** [reply]
Direct RTP streaming means you send RTP packets directly into libdatachannel's track. The streamed data must always be packetized into RTP packets, only in that case it is done by an external encoder.

**Reply by felchs** [reply]
Hi @paullouisageneau , I think I'll create a RTP stream over a VP8 packets as Unity webrtc requires vp8 or vp9 decode (https://docs.unity3d.com/Packages/com.unity.webrtc@2.4/manual/videostreaming.html)

I saw I can make vp8 packets using https://github.com/webmproject/libvpx, but I'm not sure about the RTP stream itself how can it be integrated with libdatachannel, what should I use in this case.

**Reply by paullouisageneau** [reply]
libvpx does not generate packets, only VP8-encoded frames. The second step is to packetize the VP8 data into into RTP packets (the specification to do so is [RFC 7741](https://www.rfc-editor.org/rfc/rfc7741.html)). There are different ways to proceed, for instance:
- Do the packetization  by adapting in libvpx's example [udpsample](https://chromium.googlesource.com/webm/udpsample/+/refs/heads/main).
- Do the packetization manually according to the spec with libdatachannel's `rtc::RtpHeader` helper.
- Use a higher-level multimedia framework instead of the low-level libvpx to do both the encoding and RTP packetization, like ffmpeg or gstreamer. You can then redirect the RTP output directly in a track (similar to the media example but on sender side).

Of course, you can also wait for libdatachannel to have an integrated VP8 packetizer, but I don't know when I'll have the opportunity to work on this.
